### PR TITLE
Introduce Utilities::MPI::prefix_sum

### PR DESCRIPTION
**Pull request by peterrum**
extracted from https://github.com/dealii/dealii/pull/14096

### CODE REVIEW

File: include/deal.II/base/mpi.h [hunk 1]
```diff
@@ -1370,6 +1370,33 @@ namespace Utilities
...
     compute_set_union(const std::set<T> &set, const MPI_Comm &comm);
 
+    /**
+     * Compute exclusive or incluse prefix sum for scalar quantities.
+     */
+    template <typename T>
+    std::pair<T, T>
+    prefix_sum(const T &       value,
+               const MPI_Comm &comm,
+               const bool      exclusive = true);
```
Comment (bangerth) [inline include/deal.II/base/mpi.h:1380]:
```suggestion
    /**
     * Compute exclusive or incluse prefix sum for scalar quantities.
     */
    template <typename T>
    std::pair<T, T>
    prefix_sum(const T &       value,
               const MPI_Comm &comm,
               const bool      exclusive = true);
```
The term "prefix sum" is a computer science term that is not widely understood in the broader community (for example, I have to think about it every time I see it). Would you mind explaining what a "prefix sum" is in the documentation of the function? And what "exclusive" and "inclusive" means? I think a simple formula would do wonders :-)

Separately, why does the function return a *pair* of values? Can you document what the two parts are?

File: include/deal.II/base/mpi.h [hunk 2]
```diff
@@ -2026,6 +2053,80 @@ namespace Utilities
...
+          int ierr = MPI_Scan(values.data(),
+                              temp.data(),
+                              values.size(),
+                              Utilities::MPI::mpi_type_id_for_type<T>,
+                              MPI_SUM,
+                              comm);
+          AssertThrowMPI(ierr);
+        }
+
+      Utilities::MPI::sum(values, comm, make_array_view(sums));
```
Comment (bangerth) [inline include/deal.II/base/mpi.h:2124]: Couldn't you omit this separate sum step in favor of:
* always using the `MPI_Scan`, 
* broadcasting the final value from the last process,
* if `exclusive` then subtract the value of the current process
?

**Comment by bangerth** [comment]
You will also have to guard compilation of the one function that actually does MPI communication for non-MPI compiles.

File: include/deal.II/base/mpi.h (hunk 2 above)
Comment (kronbichler) [reply include/deal.II/base/mpi.h:2124]: The difference between `Scan` and `ExScan` is that the former does
```
  \sum_{i=0}^{my_rank} values[i];
```
whereas the latter does
```
  \sum_{i=0}^{my_rank-1} values[i];
```
I however agree that we could always use the `Scan` version and subtract the local value (no need for a broadcast).

**Review by tjhei — COMMENTED** [review]
It would be great to have a wrapper around these MPI methods but I would also prefer a version that doesn't do more work than necessary.
Also, can we avoid the bool argument? I would prefer two separate functions because the bool argument always requires documenting at the call site to be understandable.

File: include/deal.II/base/mpi.h (hunk 2 above)
Comment (bangerth) [reply include/deal.II/base/mpi.h:2124]: Wait, how would you omit the broadcast? You need the value from the last process (the total sum) somehow.

**Comment by peterrum** [comment]
Let me close this PR. Splitting up the function in multiple versions does not seem to be the right way and wouldn't really in practice.

**Pull request closed without merging**